File: asg4/task.ml
(* *)
(* CSCI3180 Principles of Programming Languages *)
(* *)
(*  --- Declaration ---*)
(* *)
(*  I declare that the assignment here submitted is original except for source *)
(*  material explicitly acknowledged. I also acknowledge that I am aware of *)
(*  University policy and regulations on honesty in academic work, and of the *)
(*  disciplinary guidelines and procedures applicable to breaches of such policy *)
(*  and regulations, as contained in the website *)
(*  http://www.cuhk.edu.hk/policy/academichonesty/ *)
(* *)
(*  Assignment 4 *)	
(*  Name : Nigel Nicholas *)
(*  Student ID : 1155088791 *)
(*  Email Addr : 1155088791@link.cuhk.edu.hk *)


(*                      List.nth( list, num) 
    to access numth element of the list, starting with 0           *)
    val tasks = [
        ("t1", 2, 4),
        ("t2", 4, 6),
        ("t3", 3, 7),
        ("t4", 8, 9),
        ("t5", 1, 10),
        ("t6", 10, 1)
      ];


(*1(a)*)
(* check_task(T:string, tasklist:(string*int*int)list):bool *)
(* Recursively check from head to the head of tail whether the string matches or not *)
fun check_task(T :string, tasklist :(string*int*int)list):bool =
    
      if null tasklist then false    
        else
      let val task = hd(tasklist)
           val name = #1 task
           val startTime = #2 task
           val endTime = #3 task
      in (
        if T = name andalso startTime<=endTime then true
        else check_task(T, tl(tasklist)) 
        
        )
        

      end

(* For the purpose of returning a task  *)
fun find_task(T:string, tasklist:(string*int*int)list):(string*int*int) =
        
      let val task = hd(tasklist)
           val name = #1 task
           val startTime = #2 task
           val endTime = #3 task
      in (
        if T = name andalso startTime<endTime then task
        else find_task(T, tl(tasklist)) 
        )
        

      end


(* 1(b) *)
fun compatible(T1: string, T2: string, tasklist:(string*int*int)list):bool =
    (* check whether T1 and T2 exist, if not return false immediately, else find this 2 task *)
    if check_task(T1,tasklist) = true andalso check_task(T2,tasklist) = true then 
        let
            val task1 = find_task(T1, tasklist)
            val task2 = find_task(T2, tasklist)
            val startTime_task1 = #2 task1
            val endTime_task1 = #3 task1
            val startTime_task2 = #2 task2
            val endTime_task2 = #3 task2
        in (
            if startTime_task1 <= startTime_task2 andalso endTime_task1<=startTime_task2 then true
            else if startTime_task2 <= startTime_task1 andalso endTime_task2<=startTime_task1 then true
            else false

        )

        end
    
    else false ;


(* Check whether each task in the task list is compatible with all other task, *)
(* in imperative programming, it should be implemented with 2 for loop for i=0 ;i<n-1;i++ and for j=i+1;j<n;j++ *)

(* 1(c) *)
fun compatible_list(L: string list, tasklist: (string*int*int) list):bool =

    if null L then true
     else
    let 
        val remaining = tl L
        val head = hd L


        (* compare stable and tail *)
        (* stable is the element that does not move while doing inner loop. Used to compare this element with the next of the list *)
        fun compare(stable: string, tail: string list, tasklist: (string*int*int) list):bool = 
            if null tail then true
            else if compatible(stable, hd tail, tasklist) = true then compare(stable, tl tail, tasklist)
            else false
    
    in (
        if compare(head,remaining,tasklist) = false then false
        else compatible_list(remaining, tasklist)
    
    )
    end


  

File: asg4/task.pl
/* Insert header here */

%
% CSCI3180 Principles of Programming Languages
% 
%  --- Declaration ---
% 
%  I declare that the assignment here submitted is original except for source
%  material explicitly acknowledged. I also acknowledge that I am aware of
%  University policy and regulations on honesty in academic work, and of the
%  disciplinary guidelines and procedures applicable to breaches of such policy
%  and regulations, as contained in the website
%  http://www.cuhk.edu.hk/policy/academichonesty/
% 
%  Assignment 4	
%  Name : Nigel Nicholas    
%  Student ID : 1155088791
%  Email Addr : nigel7@cse.cuhk.edu.hk


task(t1, 2, 4).
task(t2, 4, 6).
task(t3, 3, 7).
task(t4, 8, 9).
task(t5, 1, 10).
task(t6, 10, 1).

hd([X|_],X).
hdoftl([_|[X|_]],X).
tl([_|X],X).
list_empty([], true).
list_empty([_|_], false).

notoverlap(T1,T2):-
    task(T1,_start1,_end1),
    task(T2,_start2,_end2),
    _end1 =< _start2.

/* 1a */
check_task(T):-
    task(T,_start,_end),
    _start<_end.
/* example query: check_task(t1); 
   result: yes */

/* 1b */
compatible(T1,T2):-
/* example query: compatible(t1, t2); 
   result: yes */
    check_task(T1),
    check_task(T2), 
    (notoverlap(T1,T2);notoverlap(T2,T1)).
    

/* 1c */
compatible_list(L):-
/* example query: compatible_list([t1, t2, t3]);   
   result: no */

   /* Algorithm: 
        check compatibility with one another,
        if length of the remaining tail is 0 just return true
        else if not , you have to do the recursive function of list  */
    
    length(L, _len),
    (_len>0 ->
        hd(L,_head),    
        tl(L,_tail),    
        compat(_head,_tail),
        compatible_list(_tail),
        true
    ; true 
    ).

    


compat(Head, L):-
/* compat is to check a stable card with the remaining of the list  */
    length(L,_len),
    ( _len > 0 ->
        hd(L,_header),
        tl(L,_ending),
        compatible(Head,_header),
        compat(Head,_ending),
        true
    ; true).
